File: src/main/java/net/jeikobu/jbase/config/AbstractGuildConfig.java
package net.jeikobu.jbase.config;

import net.dv8tion.jda.core.entities.Guild;

import java.util.Locale;
import java.util.Optional;

public abstract class AbstractGuildConfig {
    public AbstractGuildConfig(Guild guild) {

    }

    public abstract Optional<String> getCommandPrefix();
    public abstract void             setCommandPrefix(String prefix);

    public abstract Optional<Locale> getGuildLocale();
    public abstract void             setGuildLocale(Locale locale);

    public abstract <T> Optional<T>  getValue(String key, String defaultValue, Class<T> valueType);
    public abstract void             setValue(String key, String value);

    public <T> Optional<T> getValue(String key, Class<T> valueType) {
        return getValue(key, null, valueType);
    }
}

File: src/main/java/net/jeikobu/jbase/config/AbstractGuildConfig.kt
package net.jeikobu.jbase.config

import net.dv8tion.jda.core.entities.Guild
import org.joda.convert.StringConvert
import java.util.*
import kotlin.reflect.KClass

abstract class AbstractGuildConfig(val guild: Guild) {
    abstract var commandPrefix: String?
    abstract var guildLocale: Locale?

    inline fun <reified T : Any> setValue(key: String, value: T?) {
        if (value == null) {
            setValue(key, null)
        } else {
            setValue(key, StringConvert.INSTANCE.convertToString(value))
        }
    }

    abstract fun setValue(key: String, value: String?)

    inline fun <reified T : Any> getValue(key: String, defaultValue: String? = null): T? {
        return getValue(key, defaultValue, T::class)
    }

    abstract fun <T : Any> getValue(key: String, defaultValue: String? = null, valueType: KClass<T>): T?
}